File: 111-minimum-depth-of-binary-tree/minimum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        if (root == nullptr)
            return 0;

        // If left is NULL, go right
        if (root->left == nullptr)
            return 1 + minDepth(root->right);

        // If right is NULL, go left
        if (root->right == nullptr)
            return 1 + minDepth(root->left);

        // If both exist, take minimum
        return 1 + min(minDepth(root->left), minDepth(root->right));
    }
};

File: 111-minimum-depth-of-binary-tree/minimum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        def solve(root):
            if not root:
                return 0
            if not root.right and not root.left:
                return 1
            if not root.left:
                return solve(root.right) + 1
            if not root.right :
                return solve(root.left) + 1
            return min(solve(root.left),solve(root.right))+1
        return solve(root)